typescript/sdk/src/deploy/middleware/TokenBridgeApp.ts

```diff
@@ -1,7 +1,161 @@
-> extends HyperlaneApp<TokenBridgeContracts, Chain> {}
+> extends HyperlaneApp<TokenBridgeContracts, Chain> {
+  constructor(
+    public readonly contractsMap: ChainMap<Chain, TokenBridgeContracts>,
+    public readonly multiProvider: MultiProvider<Chain>,
+    public readonly bridgeAdapterConfigs: ChainMap<
+      Chain,
+      BridgeAdapterConfig[]
+    >,
+  ) {
+    super(contractsMap, multiProvider);
+  }
+
+  circleBridgeAdapterConfig() {
+    return objMap(
+      this.bridgeAdapterConfigs,
+      (_chain, config) =>
+        config.find(
+          (_) => _.type === BridgeAdapterType.Circle,
+        ) as CircleBridgeAdapterConfig,
+    );
+  }
+
+  async fetchCircleMessageTransactions(chain: Chain): Promise<string[]> {
+    const cc = this.multiProvider.getChainConnection(chain);
+    const params = new URLSearchParams({
+      module: 'logs',
+      action: 'getLogs',
+      address: this.getContracts(chain).circleBridgeAdapter!.address,
+      topic0: CircleBridgeAdapterInterface.getEventTopic(
+        CircleBridgeAdapterInterface.getEvent('BridgedToken'),
+      ),
+    });
+    const req = await fetch(`${cc.getApiUrl()}?${params}`);
+    const response = await req.json();
+
+    return response.result.map((_: any) => _.transactionHash).flat();
+  }
+
+  async parseCircleMessages(
+    chain: Chain,
+    txHash: string,
+  ): Promise<CircleBridgeMessage<Chain>[]> {
+    const connection = this.multiProvider.getChainConnection(chain);
+    const receipt = await connection.provider.getTransactionReceipt(txHash);
+    const matchingLogs = receipt.logs
+      .map((_) => {
+        try {
+          return [CircleBridgeInterface.parseLog(_)];
+        } catch {
+          try {
+            return [CircleBridgeAdapterInterface.parseLog(_)];
+          } catch {
+            return [];
+          }
+        }
+      })
+      .flat();
+
+    if (matchingLogs.length == 0) return [];
+    const message = matchingLogs.find((_) => _!.name === 'MessageSent')!.args
+      .message;
+    const nonce = matchingLogs.find((_) => _!.name === 'BridgedToken')!.args
+      .nonce;
+    const remoteChain =
+      message.chain === Chains.fuji ? Chains.goerli : Chains.fuji;
+    return [
+      {
+        chain,
+        // @ts-ignore
+        remoteChain,
+        txHash,
+        message,
+        nonce,
+        domain: 0,
+        nonceHash: ethers.utils.solidityKeccak256(
+          ['uint32', 'uint256'],
+          [0, nonce],
+        ),
+      },
+    ];
+  }
+
+  async attemptCircleAttestationSubmission(
+    message: CircleBridgeMessage<Chain>,
+  ) {
+    const connection = this.multiProvider.getChainConnection(
+      message.remoteChain,
+    );
+    const transmitter = ICircleMessageTransmitter__factory.connect(
+      this.circleBridgeAdapterConfig()[message.remoteChain]
+        .messageTransmitterAddress,
+      connection.signer!,
+    );
+
+    const alreadyProcessed = await transmitter.usedNonces(message.nonceHash);
```

Comment:
ditto
is there no way to filter for this using logs

Reply:
What exactly do you mean? Yeah I mean you could make another request to Etherscan for completion logs, but this seems preferrable imo. Requests to RPC nodes are pretty easy